Add add_bits parameter to QuantumCircuit.measure_all

dani-guijo opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
I added the add_creg boolean parameter to the measure_all() method to fix issue #4158. If add_creg=True, the code will behave as it used to, creating new classical register to store the measurement results. If set to False, it will measure the qubits in the existing classical register (if it's big enough) or raise an error (if it's not big enough). It has been tested using tox and the documentation as been updated accordingly.


### Details and comments

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=7305) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=7305) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit-terra?pullRequest=7305) it.</sub>

dani-guijo commented:
Apparently I added two commits accidentaly. Only the second one (the one I updated later) is the one I wanted to submit. Is there anything I should do?

dani-guijo commented:
Looks like two of the checks failed. It worked well on my laptop so I guess the changes needed are minimal. Please tell me if there's anything else I have to do.

enavarro51 commented:
Just do `black qiskit test tools` and recommit and that should take care of it.

jakelishman reviewed (commented):
Edwin is right, the regular tests are passing, but we have code-style tests as well, and that's what's going wrong here.  You can run the same code-style tests by running `tox -e lint` assuming you've followed the contributing guide for getting your development environment (they can take about 15 minutes and not look like they're doing anything, unfortunately).  Here it's likely only formatting problems, and Edwin's `black` command will fix all of them for you automatically.

For the PR, please could you also add some tests of the new behaviour (there are some `measure_all` tests in `test/python/circuit/test_circuit_operations.py`), and add a release note about the new behaviour?

dani-guijo commented:
I already run the 'black qiskit test tools' command and it ran smoothly in my laptop, but it looks like it's still failling. Any guesses?
Also, I haven't yet added the new tests to the documentation (I intend to), but it should work with the existing ones.

enavarro51 commented:
Hopefully you can follow me here. Try this.
- On the second red 'x' in the list of checks (the one with 'Lint_Docs...') click on 'Details'
- In the middle of the next screen, click on the blue '1 errors / 0 warnings'
- In the next screen, click on the red circle 'x' that says 'Lint Python 3.7'
- Finally, click on the red circle 'x' that says 'Style and ...'
- You should now see your 2 lint errors

@jakelishman is right that a full 'tox -e lint' would have shown these. It takes a while to run locally, so I usually try to fix 'black' errors first since that catches most of the lint.

mtreinish commented:
@enavarro51's instructions to find the error message from CI will show you what is wrong. You can also run `tox -elint-incr` to run a subset of the lint checks locally for a faster result. It's not quite as exhaustive but it will get you like 95% (it mainly misses some of the import cycle checks) of the coverage that the CI job is running. The style and lint checks usage are documented here: https://github.com/Qiskit/qiskit-terra/blob/main/CONTRIBUTING.md#style-and-lint

dani-guijo commented:
Thaks @enavarro51 and @mtreinish for your advice, it was a lot easier this way. Every check is successful now and I'm waiting for a mantainer to approve it.

This is my first contribution to Qiskit so thank you so much for the patience!

jakelishman commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -2205,27 +2205,41 @@ def measure_active(self, inplace: bool = True) -> Optional["QuantumCircuit"]:
 
-    def measure_all(self, inplace: bool = True) -> Optional["QuantumCircuit"]:
+    def measure_all(
+        self, inplace: bool = True, add_creg: bool = True
```
I should have thought about this before, but please could you change the name of the argument to `add_bits`?  Nowadays, the bits are the important part, and the register is just a handle to a collection of bits, so it's more important that we're adding bits.

jakelishman commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -2205,27 +2205,41 @@ def measure_active(self, inplace: bool = True) -> Optional["QuantumCircuit"]:
+        self, inplace: bool = True, add_creg: bool = True
+    ) -> Optional["QuantumCircuit"]:
         """Adds measurement to all qubits. Creates a new ClassicalRegister with a
-        size equal to the number of qubits being measured.
+        size equal to the number of qubits being measured if `add_creg = True`.
```
It's good that you updated this.  We should be a bit more descriptive about what happens if you set `add_bits=False` as well - something like

> Add a measurement to all qubits.
>
> By default, a register of new classical bits is added to the circuit to store these measurements.  If the argument ` ``add_bits=False`` `, then the results will be stored in the classical bits that are already present, with qubit ` ``n`` ` being measured into classical bit ` ``n`` `.

(as an aside: in Sphinx, if you just want monospaced text, you need to use two backticks (` ``monospaced`` `), not just one)

jakelishman commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -2205,27 +2205,41 @@ def measure_active(self, inplace: bool = True) -> Optional["QuantumCircuit"]:
 
         Args:
             inplace (bool): All measurements inplace or return new circuit.
+            add_creg (bool): Adds a new classical register.
```
```suggestion
            add_bits (bool): Whether to add new bits to store the results.
```

jakelishman commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -2205,27 +2205,41 @@ def measure_active(self, inplace: bool = True) -> Optional["QuantumCircuit"]:
         Returns:
             QuantumCircuit: Returns circuit with measurements when `inplace = False`.
+
+        Raises:
+            CircuitError: if `add_creg = False` and `len(circ.clbits) < len(circ.qubits)`.
```
If possible, it's better to describe the condition in words rather than code, because it's easier for people to understand, usually.  So perhaps we could say something like

```python
Raises:
    CircuitError: if ``add_bits=False`` but there are not enough classical bits in the circuit.
```

jakelishman commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -2205,27 +2205,41 @@ def measure_active(self, inplace: bool = True) -> Optional["QuantumCircuit"]:
+        else:
+            if len(circ.clbits) < len(circ.qubits):
+                raise CircuitError(
+                    "The size of the ClassicalRegister must be equal or greater than"
+                    "the size of the QuantumRegister."
```
It's more that the number of classical bits must be greater than or equal to the number of qubits - it's possible to have more than one (or even zero!) registers in the circuit.

(Also, there will be a missing space in the message between "than" and "the".)

jakelishman commented on test/python/circuit/test_circuit_operations.py:
```diff
@@ -440,6 +440,22 @@ def test_measure_all(self):
 
         self.assertEqual(expected, circuit)
 
+    def test_measure_all_not_add_creg(self):
+        """Test measure_all applies measurements to all qubits.
+        Does not create a new ClassicalRegister if the existing one is big enough.
+        """
+        qr = QuantumRegister(2)
+        cr = ClassicalRegister(2, "meas")
+
+        circuit = QuantumCircuit(qr, cr)
+        circuit.measure_all(add_creg=False)
+
+        expected = QuantumCircuit(qr, cr)
+        expected.barrier()
+        expected.measure(qr, cr)
+
+        self.assertEqual(expected, circuit)
+
```
This is a good test - it's very clear.

Please could you also add a test that it works if there are more classical bits than qubits, and a test that the correct error is raised if you try to use `add_bits=False` without enough classical bits?

jakelishman reviewed (commented):
This is looking good, thanks!

I've left some comments about some of the documentation (which you may not have known), but the code itself looks good.

Please don't forget to add a "feature" release note - if you've got the dev requirements installed, you should be able to run `reno new measure_all-add_bits` to get the skeleton file generated for you.

dani-guijo commented:
This is great input, I'll get on it right away.

jakelishman commented on test/python/circuit/test_circuit_operations.py:
```diff
@@ -480,9 +480,8 @@ def test_measure_all_not_add_bits_smaller(self):
-        circuit.measure_all(add_bits=False)
 
-        self.assertRaises(CircuitError, circuit)
+        self.assertRaises(CircuitError, circuit.measure_all(add_bits=False))
```
Here you're looking for something like
```python
with self.assertRaisesRegex(CircuitError, "<a bit of the message you expect>"):
    circuit.measure_all(add_bits=False)
```

dani-guijo commented:
I think everything is solved now.

dani-guijo commented:
I forgot to add the release note, the rest is still the same.

jakelishman reviewed (approved):
I pushed a minor commit (c2f63979a) to fix up a couple of bits of Sphinx formatting and put some links into the release note, but this looks good to merge.

Thank you for doing this, and for your patience in getting it in!

dani-guijo commented:
That's great! Thanks @jakelishman and everyone else for the input, I learned a lot doing this and I'll keep contributing whenever I have the time.

jakelishman commented:
No problem, we'll be glad to have you!  This should merge later today, and it'll be live in the released version of Terra sometime within the next couple of weeks.
